File: sync-bridge-rust/src/dto.rs
// src/dto.rs
use crate::errors::AppError;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// --- Helper Functions for Validation ---
fn is_blank(s: &str) -> bool {
    s.trim().is_empty()
}

fn is_valid_email(s: &str) -> bool {
    // A simple robust email regex check or character check
    s.contains('@')
        && s.split('@')
            .nth(1)
            .is_some_and(|domain| domain.contains('.'))
}

// --- Sync Request ---
#[derive(Debug, Deserialize)]
pub struct SyncRequest {
    pub model: String,
    pub data: Vec<serde_json::Value>,
}

impl SyncRequest {
    pub fn validate(&self) -> Result<(), AppError> {
        let mut errors = HashMap::new();
        if is_blank(&self.model) {
            errors.insert("model".to_string(), "must not be blank".to_string());
        } else if !["customers", "products", "orders", "employees"].contains(&self.model.as_str()) {
            errors.insert("model".to_string(), "Invalid model".to_string());
        }

        if self.data.is_empty() {
            errors.insert("data".to_string(), "must not be empty".to_string());
        }

        if !errors.is_empty() {
            return Err(AppError::Validation { errors });
        }
        Ok(())
    }
}

// --- Customer ---
#[derive(Debug, Deserialize, Serialize)]
pub struct CustomerDto {
    pub id: Option<i64>,
    pub email: String,
    #[serde(rename = "first_name")]
    pub first_name: String,
    #[serde(rename = "last_name")]
    pub last_name: String,
    #[serde(rename = "default_currency")]
    pub default_currency: Option<String>,
}

impl CustomerDto {
    pub fn validate(&self) -> Result<(), AppError> {
        let mut errors = HashMap::new();
        if is_blank(&self.email) {
            errors.insert("email".to_string(), "must not be blank".to_string());
        } else if !is_valid_email(&self.email) {
            errors.insert(
                "email".to_string(),
                "must be a well-formed email address".to_string(),
            );
        }

        if is_blank(&self.first_name) {
            errors.insert("first_name".to_string(), "must not be blank".to_string());
        }

        if is_blank(&self.last_name) {
            errors.insert("last_name".to_string(), "must not be blank".to_string());
        }

        if let Some(ref currency) = self.default_currency {
            if currency.len() != 3 {
                errors.insert("default_currency".to_string(), "size must be 3".to_string());
            }
        }

        if !errors.is_empty() {
            return Err(AppError::Validation { errors });
        }
        Ok(())
    }
}

// --- Product ---
#[derive(Debug, Deserialize, Serialize)]
pub struct ProductDto {
    pub id: Option<i64>,
    pub name: String,
    pub description: Option<String>,
    pub price: i32,
    pub currency: Option<String>,
    pub active: Option<bool>,
    #[serde(rename = "weight_grams")]
    pub weight_grams: Option<i32>,
}

impl ProductDto {
    pub fn validate(&self) -> Result<(), AppError> {
        let mut errors = HashMap::new();
        if is_blank(&self.name) {
            errors.insert("name".to_string(), "must not be blank".to_string());
        }

        if let Some(ref currency) = self.currency {
            if currency.len() != 3 {
                errors.insert("currency".to_string(), "size must be 3".to_string());
            }
        }

        if !errors.is_empty() {
            return Err(AppError::Validation { errors });
        }
        Ok(())
    }
}

// --- Order Item ---
#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct OrderItemDto {
    pub id: Option<i64>,
    #[serde(rename = "product_id")]
    pub product_id: i64,
    pub qty: i32,
    #[serde(rename = "unit_price")]
    pub unit_price: i32,
}

impl OrderItemDto {
    pub fn validate(&self) -> Result<(), AppError> {
        let mut errors = HashMap::new();
        if self.product_id <= 0 {
            errors.insert("product_id".to_string(), "must be positive".to_string());
        }
        if self.qty <= 0 {
            errors.insert("qty".to_string(), "must be greater than 0".to_string());
        }
        if self.unit_price <= 0 {
            errors.insert("unit_price".to_string(), "must be positive".to_string());
        }

        if !errors.is_empty() {
            return Err(AppError::Validation { errors });
        }
        Ok(())
    }
}

// --- Order ---
#[derive(Debug, Deserialize, Serialize)]
pub struct OrderDto {
    pub id: Option<i64>,
    #[serde(rename = "order_number")]
    pub order_number: String,
    #[serde(rename = "customer_id")]
    pub customer_id: i64,
    pub status: String,
    pub currency: Option<String>,
    pub amount: Option<i32>,
    pub items: Option<Vec<OrderItemDto>>,
}

impl OrderDto {
    pub fn validate(&self) -> Result<(), AppError> {
        let mut errors = HashMap::new();
        if is_blank(&self.order_number) {
            errors.insert("order_number".to_string(), "must not be blank".to_string());
        }

        if self.customer_id <= 0 {
            errors.insert("customer_id".to_string(), "must be positive".to_string());
        }

        let valid_statuses = [
            "pending",
            "paid",
            "shipped",
            "completed",
            "cancelled",
            "refunded",
        ];
        if is_blank(&self.status) {
            errors.insert("status".to_string(), "must not be blank".to_string());
        } else if !valid_statuses.contains(&self.status.as_str()) {
            errors.insert(
                "status".to_string(),
                "must match pending|paid|shipped|completed|cancelled|refunded".to_string(),
            );
        }

        if let Some(ref currency) = self.currency {
            if currency.len() != 3 {
                errors.insert("currency".to_string(), "size must be 3".to_string());
            }
        }

        if !errors.is_empty() {
            return Err(AppError::Validation { errors });
        }
        Ok(())
    }
}

// --- Employee ---
#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct EmployeeDto {
    pub id: String,
    #[serde(rename = "employeeId")]
    pub employee_id: String,
    #[serde(rename = "firstName")]
    pub first_name: String,
    #[serde(rename = "middleName")]
    pub middle_name: Option<String>,
    #[serde(rename = "lastName")]
    pub last_name: String,
    pub gender: Option<String>,
    pub email: String,
    #[serde(rename = "phoneNumber")]
    pub phone_number: Option<String>,
    #[serde(rename = "dateOfBirth")]
    pub date_of_birth: Option<DateTime<Utc>>,
    pub nationality: Option<String>,
    #[serde(rename = "jobLevel")]
    pub job_level: Option<String>,
    pub department: Option<String>,
    pub location: Option<String>,
    #[serde(rename = "bankAccountNumber")]
    pub bank_account_number: Option<String>,
    pub company: Option<String>,
    #[serde(rename = "jobTitle")]
    pub job_title: Option<String>,
    #[serde(rename = "costCenter")]
    pub cost_center: Option<String>,
    #[serde(rename = "startDate")]
    pub start_date: Option<DateTime<Utc>>,
    #[serde(rename = "employeeStatus")]
    pub employee_status: Option<String>,
    #[serde(rename = "managerId")]
    pub manager_id: Option<String>,
    #[serde(rename = "managerEmail")]
    pub manager_email: Option<String>,
    #[serde(rename = "lastModifiedOn")]
    pub last_modified_on: Option<DateTime<Utc>>,
    #[serde(rename = "lastModified")]
    pub last_modified: Option<i64>,
}

impl EmployeeDto {
    pub fn validate(&self) -> Result<(), AppError> {
        let mut errors = HashMap::new();
        if is_blank(&self.id) {
            errors.insert("id".to_string(), "must not be blank".to_string());
        }
        if is_blank(&self.employee_id) {
            errors.insert("employeeId".to_string(), "must not be blank".to_string());
        }
        if is_blank(&self.first_name) {
            errors.insert("firstName".to_string(), "must not be blank".to_string());
        }
        if is_blank(&self.last_name) {
            errors.insert("lastName".to_string(), "must not be blank".to_string());
        }
        if is_blank(&self.email) {
            errors.insert("email".to_string(), "must not be blank".to_string());
        } else if !is_valid_email(&self.email) {
            errors.insert(
                "email".to_string(),
                "must be a well-formed email address".to_string(),
            );
        }

        if let Some(ref m_email) = self.manager_email {
            if !is_blank(m_email) && !is_valid_email(m_email) {
                errors.insert(
                    "managerEmail".to_string(),
                    "must be a well-formed email address".to_string(),
                );
            }
        }

        if !errors.is_empty() {
            return Err(AppError::Validation { errors });
        }
        Ok(())
    }
}

File: sync-bridge-go/internal/dto/dto.go
package dto

import (
	"encoding/json"

	"github.com/go-playground/validator/v10"
)

var validate = validator.New()

type SyncRequest struct {
	Model string            `json:"model" validate:"required,oneof=customers products orders employees"`
	Data  []json.RawMessage `json:"data" validate:"required,min=1" swaggertype:"array,object"`
}

func (s *SyncRequest) Validate() error {
	return validate.Struct(s)
}

type CustomerDto struct {
	ID              *int64  `json:"id"`
	Email           string  `json:"email" validate:"required,email"`
	FirstName       string  `json:"first_name" validate:"required"`
	LastName        string  `json:"last_name" validate:"required"`
	DefaultCurrency *string `json:"default_currency" validate:"omitempty,len=3"`
}

func (c *CustomerDto) Validate() error {
	return validate.Struct(c)
}

type ProductDto struct {
	ID          *int64  `json:"id"`
	Name        string  `json:"name" validate:"required"`
	Description *string `json:"description"`
	Price       *int    `json:"price" validate:"required"`
	Currency    *string `json:"currency" validate:"omitempty,len=3"`
	Active      *bool   `json:"active"`
	WeightGrams *int    `json:"weight_grams"`
}

func (p *ProductDto) Validate() error {
	return validate.Struct(p)
}

type OrderItemDto struct {
	ID        *int64 `json:"id"`
	ProductID *int64 `json:"product_id" validate:"required"`
	Qty       *int   `json:"qty" validate:"required"`
	UnitPrice *int   `json:"unit_price" validate:"required"`
}

func (o *OrderItemDto) Validate() error {
	return validate.Struct(o)
}

type OrderDto struct {
	ID          *int64          `json:"id"`
	OrderNumber string          `json:"order_number" validate:"required"`
	CustomerID  *int64          `json:"customer_id" validate:"required"`
	Status      string          `json:"status" validate:"required,oneof=pending paid shipped completed cancelled refunded"`
	Currency    *string         `json:"currency" validate:"omitempty,len=3"`
	Amount      *int            `json:"amount"`
	Items       *[]OrderItemDto `json:"items"`
}

func (o *OrderDto) Validate() error {
	return validate.Struct(o)
}

type EmployeeDto struct {
	ID                string  `json:"id" validate:"required"`
	EmployeeID        string  `json:"employeeId" validate:"required"`
	FirstName         string  `json:"firstName" validate:"required"`
	MiddleName        *string `json:"middleName"`
	LastName          string  `json:"lastName" validate:"required"`
	Gender            *string `json:"gender"`
	Email             string  `json:"email" validate:"required,email"`
	PhoneNumber       *string `json:"phoneNumber"`
	DateOfBirth       *string `json:"dateOfBirth"`
	Nationality       *string `json:"nationality"`
	JobLevel          *string `json:"jobLevel"`
	Department        *string `json:"department"`
	Location          *string `json:"location"`
	BankAccountNumber *string `json:"bankAccountNumber"`
	Company           *string `json:"company"`
	JobTitle          *string `json:"jobTitle"`
	CostCenter        *string `json:"costCenter"`
	StartDate         *string `json:"startDate"`
	EmployeeStatus    *string `json:"employeeStatus"`
	ManagerID         *string `json:"managerId"`
	ManagerEmail      *string `json:"managerEmail" validate:"omitempty,email"`
	LastModifiedOn    *string `json:"lastModifiedOn"`
	LastModified      *int64  `json:"lastModified"`
}

func (e *EmployeeDto) Validate() error {
	return validate.Struct(e)
}
